src/thingdef/thingdef_codeptr.cpp

```diff
@@ -6003,12 +6021,35 @@ DEFINE_ACTION_FUNCTION_PARAMS(AActor, A_CheckProximity)
 		else if (classname != mo->GetClass())
 			continue;
 
-		//Make sure it's in range and respect the desire for Z or not.
-		if (ref->AproxDistance(mo) < distance &&
+		//[MC]Make sure it's in range and respect the desire for Z or not. The function forces it to use
+		//Z later for ensuring CLOSEST and FARTHEST flags are respected perfectly.
+		//Ripped from sphere checking in A_RadiusGive (along with a number of things).
+		if ((flags & CPXF_NODISTANCE) || (ref->AproxDistance(mo) < distance &&
 			((flags & CPXF_NOZ) ||
-			((ref->Z() > mo->Z() && ref->Top() < distance) ||
-			(ref->Z() <= mo->Z() && mo->Z() - ref->Top() < distance))))
+			((ref->Z() > mo->Z() && ref->Z() - mo->Top() < distance) ||
+			(ref->Z() <= mo->Z() && mo->Z() - ref->Top() < distance)))))
 		{
+			if ((flags & CPXF_CHECKSIGHT) && !(P_CheckSight(mo, ref, SF_IGNOREVISIBILITY | SF_IGNOREWATERBOUNDARY)))
+				continue;
+
+			if (ptrWillChange)
+			{
+				current = ref->AproxDistance(mo);
+
+				if ((flags & CPXF_CLOSEST) && ((current < closer) || !closer))
```

Comment:
I was trying to debug something with it but forgot to remove it afterwards. It's gone now.